src/anndata/_core/merge.py

```diff
@@ -647,21 +750,56 @@ def _apply_to_cupy_array(self, el, *, axis, fill_value=None):
 
         return out
 
-    def _apply_to_array(self, el, *, axis, fill_value=None):
+    def _apply_to_array_api(self, el, *, axis, fill_value=None):
         if fill_value is None:
             fill_value = default_fill_value([el])
-        if el.shape[axis] == 0:
-            # Presumably faster since it won't allocate the full array
-            shape = list(el.shape)
-            shape[axis] = len(self.new_idx)
-            return np.broadcast_to(fill_value, tuple(shape))
 
         indexer = self.idx
 
-        # Indexes real fast, and does outer indexing
-        return pd.api.extensions.take(
-            el, indexer, axis=axis, allow_fill=True, fill_value=fill_value
-        )
+        if _is_pandas(el):
+            # using the behavior that already exists in pandas for Series/Index
+            return pd.api.extensions.take(
+                el, indexer, axis=axis, allow_fill=True, fill_value=fill_value
+            )
+        # e.g., numpy, jax.numpy, cubed, etc.
+        xp = get_namespace(el)
+
+        # Handling edge case to mimic pandas behavior
+        if el.shape[axis] == 0:
+            shape = list(el.shape)
+            shape[axis] = indexer.shape[0]
+            # convert fill_value to the same type as el - to keep everything in the same dtype
+            # fv = xp.asarray(fill_value, dtype=getattr(el, "dtype", None))
+            fv = xp.asarray(fill_value)
+            return xp.broadcast_to(fv, shape)
+
+        # Check: is array-api compatible, but not NumPy
+        if not isinstance(el, np.ndarray) and _is_array_api_compatible(el):
+            # Convert to NumPy via DLPack
+            el_np = _dlpack_to_numpy(el)
+
+            # Recursively call this same function
+            out_np = self._apply_to_array_api(el_np, axis=axis, fill_value=fill_value)
+
+            # reverting back to numpy as it is hard to reindex on JAX and others
+            return _dlpack_from_numpy(out_np, xp)
+
+        # numpy case
```

Comment:
I think the logic here is a little confused.  This function used to be for `numpy`, but you added the `_is_pandas` check above, which I don't think applies here.  But the logic you've written "# numpy case" and down works great for non-numpy array-api compatible arrays as well!

So I would leave the numpy case as before (i.e., remove `_is_pandas` and check `isinstance(el, np.ndarray)`, and then in the case it is not a numpy array, use this logic under "# numpy case"! You can then get rid of the `if not isinstance(el, np.ndarray) and _is_array_api_compatible(el):` branch.  One reason I cautioned against falling back to `numpy` behavior is that some things like `jax` arrays that are API compatible might be on the GPU! You can't transfer a JAX array on the GPU to numpy :/